contrib/smtp_proxy/filters/network/source/smtp_decoder.cc

```diff
@@ -0,0 +1,299 @@
+#include "contrib/smtp_proxy/filters/network/source/smtp_decoder.h"
+
+#include "source/common/common/logger.h"
+
+#include "absl/strings/match.h"
+#include "contrib/smtp_proxy/filters/network/source/smtp_utils.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace NetworkFilters {
+namespace SmtpProxy {
+
+Decoder::Result DecoderImpl::onData(Buffer::Instance& data, bool upstream) {
+  Decoder::Result result = Decoder::Result::ReadyForNext;
+
+  if (upstream) {
+    result = parseResponse(data);
+    data.drain(data.length());
+    return result;
+  }
+  result = parseCommand(data);
+  data.drain(data.length());
+  return result;
+}
+
+Decoder::Result DecoderImpl::parseCommand(Buffer::Instance& data) {
+  ENVOY_LOG(debug, "smtp_proxy: decoding {} bytes", data.length());
+  Decoder::Result result = Decoder::Result::ReadyForNext;
+  std::string command = data.toString();
+
+  if (command.length() < 6) {
+    // Message size is not sufficient to parse.
+    return result;
+  }
```

Comment:
If the command length is less then 6, then the parseCommand will do nothing return deirectly. And then the `onData` call will drain the unparsed buffer? I think here are some error, or I miss something?

Reply:
If command length is <6, currently the smtp filter will pass it to next filter (tcp filter) and it will reach upstream. Upstream can reject the command with 4xx bad syntax. But we can have smtp filter itself reject commands with length < 6. Will do this change

Reply:
> If command length is <6, currently the smtp filter will pass it to next filter (tcp filter) and it will reach upstream. Upstream can reject the command with 4xx bad syntax. But we can have smtp filter itself reject commands with length < 6. Will do this change

Do not do it. Upstream server has its own buffering. The logic behind buffering partial packets is to address the fact that packets may arrive to the filter in unknown chunks. For example, it filter receives 3 bytes, pass those 3 bytes upstream and continue buffering until filter has a protocol unit which can be parsed. In the meantime, keep passing upstream whatever arrives from downstream.
In postgres I used logic to wait for minimum number of bytes, because that represented postgres header and each packet has that header. Not sure about smtp.

Reply:
Hey @cpakulski ,
I refactored the logic to parse smtp commands and responses. Please have a look at DecoderImpl::parseCommand() function.

Reply:
I didn't think @cpakulski 's (and my) concern is be addressed. From the source code, the input buffer will always be drained even if there is no complete command and nothing is processed accutally.
Or I missed something?

/wait

Reply:
@wbpcode Got you point. I am doing data.drain() each time e.g.
  `result = parseCommand(data);
  data.drain(data.length());`
  Will make changes to buffer the received bytes if they are not sufficient to parse.